ports/posix/iot_logging.c

```diff
@@ -0,0 +1,405 @@
+
+/*-----------------------------------------------------------*/
+
+/**
+ * @brief Lookup table for log levels.
+ *
+ * Converts one of the @ref logging_constants_levels to a string.
+ */
+static const char * const _pLogLevelStrings[ 5 ] =
+{
+    "",      /* IOT_LOG_NONE */
+    "ERROR", /* IOT_LOG_ERROR */
+    "WARN ", /* IOT_LOG_WARN */
+    "INFO ", /* IOT_LOG_INFO */
+    "DEBUG"  /* IOT_LOG_DEBUG */
+};
+
+/*-----------------------------------------------------------*/
+
+#if !defined( IOT_STATIC_MEMORY_ONLY ) || ( IOT_STATIC_MEMORY_ONLY == 0 )
+    static bool _reallocLoggingBuffer( void ** pOldBuffer,
+                                       size_t newSize,
+                                       size_t oldSize )
+    {
+        bool status = false;
+
+        /* Allocate a new, larger buffer. */
+        void * pNewBuffer = IotLogging_Malloc( newSize );
+
+        /* Ensure that memory allocation succeeded. */
+        if( pNewBuffer != NULL )
+        {
+            /* Copy the data from the old buffer to the new buffer. */
+            ( void ) memcpy( pNewBuffer, *pOldBuffer, oldSize );
+
+            /* Free the old buffer and update the pointer. */
+            IotLogging_Free( *pOldBuffer );
+            *pOldBuffer = pNewBuffer;
+
+            status = true;
+        }
+
+        return status;
+    }
+#endif /* if !defined( IOT_STATIC_MEMORY_ONLY ) || ( IOT_STATIC_MEMORY_ONLY == 0 ) */
+
+/*-----------------------------------------------------------*/
+
+void IotLog_Generic( int32_t messageLevel,
+                     const char * const pFormat,
+                     ... )
+{
+    int requiredMessageSize = 0;
+    size_t bufferSize = 0,
+           bufferPosition = 0, timestringLength = 0;
+    char * pLoggingBuffer = NULL;
+    va_list args;
+
+    if( messageLevel == 0 )
+    {
+        return;
+    }
+
+    /* Add length of log level if requested. */
+    bufferSize += MAX_LOG_LEVEL_LENGTH;
+
+    /* Add length of timestring. */
+    bufferSize += MAX_TIMESTRING_LENGTH;
+
+    /* Add 64 as an initial (arbitrary) guess for the length of the message. */
+    bufferSize += 64;
+
+    /* In static memory mode, check that the log message will fit in the a
+     * static buffer. */
+    #if IOT_STATIC_MEMORY_ONLY == 1
+        if( bufferSize >= IotLogging_StaticBufferSize() )
+        {
+            /* If the static buffers are likely too small to fit the log message,
+             * return. */
+            return;
+        }
+
+        /* Otherwise, update the buffer size to the size of a static buffer. */
+        bufferSize = IotLogging_StaticBufferSize();
+    #endif
+
+    /* Allocate memory for the logging buffer. */
+    pLoggingBuffer = ( char * ) IotLogging_Malloc( bufferSize );
+
+    if( pLoggingBuffer == NULL )
+    {
+        return;
+    }
+
+    /* Print the message log level. */
+    /* Ensure that message level is valid. */
+    if( ( messageLevel >= IOT_LOG_NONE ) && ( messageLevel <= IOT_LOG_DEBUG ) )
```

Comment:
The only time messageLevel  can be invalid is when someone calls this function directly - which is more of misuse. I'd suggest to put an assert at the top.

Reply:
Removed the if-check and added an assert check.